### PR TITLE
Remove goroutines from limiter middleware.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
The `limitedRoundTripper` would spawn thousands if not millions of goroutines. However, they were mostly idle waiting for downstream calls. This would congest the Go scheduler.

A semaphore is more suitable to control concurrency as the [example](https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool) in the documentation illustrates.

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/querier/queryrange/limits.go [hunk 1]
```diff
@@ -528,40 +507,30 @@ func (rt limitedRoundTripper) RoundTrip(r *http.Request) (*http.Response, error)
...
+			defer sem.Release(int64(1))
+
+			return rt.do(ctx, r)
 		})).Do(ctx, request)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO(karsten): Do we have to wait for all work being done
+	// sem.Acquire(ctx, int64(parallelism))
```
Comment (jeschkies) [inline pkg/querier/queryrange/limits.go:533]: I don't think so since we are waiting in line 512 but we could just to be safe.

Comment (chaudum) [reply pkg/querier/queryrange/limits.go:533]: Agree, it does not seem to be necessary.

**Review by chaudum — APPROVED** [review]
Nice find!

**Review by dannykopping — APPROVED** [review]
🚀🚀🚀

**Pull request merged**